### PR TITLE
[BEAM-3876] avoid NPE if checkpoint is null in an unbounded source

### PR DESCRIPTION
**Pull request by rmannibucau**
when there is no element the unbounded source can have no checkpoint at all so it shouldnt be finalized in such a case

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by jbonofre** [comment]
Waiting the end of Jenkins to merge.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java [hunk 1]
```diff
@@ -171,10 +171,28 @@ public void processElement(
...
+                UnboundedReader<?> toClose = reader;
+                reader = null; // Avoid double close below in case of an exception.
+                toClose.close();
+              } catch (final IOException closeEx) {
+                if (ioe != null) {
+                  ioe.addSuppressed(closeEx);
+                } else {
+                  throw closeEx;
+                }
+              }
```
Comment (lukecwik) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java]: Your forgetting to throw `ioe` if close doesn't throw.

Please add a test to cover these cases.

Comment (rmannibucau) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java]: good catch, updated, thanks!

**Comment by jbonofre** [comment]
retest this please

**Comment by jbonofre** [comment]
retest this please

**Comment by rmannibucau** [comment]
retest this please

**Comment by lukecwik** [comment]
retest this please

**Pull request closed without merging**